src/lib.rs

```diff
@@ -7,28 +7,30 @@
 extern crate test;
 
 /// An iterator that runs all at once
-pub trait Traversal<T>: Sized {
+pub trait Traversal: Sized {
+    type Item;
+
     /// Run this Iterator using the provided closure.
     ///
     /// Return false from the closure to end the iteration.
-    fn foreach<F: FnMut(T) -> bool>(self, F);
+    fn foreach<F>(self, F) where F: FnMut(Self::Item) -> bool;
 
     /// Run this Iterator using the provided closure.
     ///
     /// This is a utility method for non-cancelling iterations.
-    fn run<F: FnMut(T)>(self, mut f: F) {
-        self.foreach(|&mut: t: T| { f(t); false })
+    fn run<F: FnMut(Self::Item)>(self, mut f: F) {
```

Comment:
It's not a big deal, but it seems like the right choice for consistency? The meat of the method is that it's `fn<F>(self, mut f: F)`.

Your call.